src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs

```diff
@@ -838,6 +839,125 @@ public async Task UnrecognizedMediaType()
         Assert.Contains(TestSink.Writes, w => w.Message.Contains("Unrecognized Content-Type for body."));
     }
 
+    [Fact]
+    public async Task UpgradeToWebSocketLogsResponsePropertiesAndHeadersWhenResponseIsFlushed()
+    {
+        var options = CreateOptionsAccessor();
+        options.CurrentValue.LoggingFields = HttpLoggingFields.ResponsePropertiesAndHeaders;
+
+        var writtenHeaders = new TaskCompletionSource();
+        var letBodyFinish = new TaskCompletionSource();
+
+        var httpContext = new DefaultHttpContext();
+
+        var upgradeFeatureMock = new Mock<IHttpUpgradeFeature>();
+        upgradeFeatureMock.Setup(m => m.IsUpgradableRequest).Returns(true);
+        upgradeFeatureMock
+            .Setup(m => m.UpgradeAsync())
+            .Callback(() =>
+            {
+                httpContext.Response.StatusCode = StatusCodes.Status101SwitchingProtocols;
+                httpContext.Response.Headers[HeaderNames.Connection] = HeaderNames.Upgrade;
+            })
+            .ReturnsAsync(Stream.Null);
+        httpContext.Features.Set<IHttpUpgradeFeature>(upgradeFeatureMock.Object);
+
+        var middleware = new HttpLoggingMiddleware(
+            async c =>
+            {
+                await c.Features.Get<IHttpUpgradeFeature>().UpgradeAsync();
+                await c.Response.CompleteAsync();
+                writtenHeaders.SetResult();
+                await letBodyFinish.Task;
+            },
+            options,
+            LoggerFactory.CreateLogger<HttpLoggingMiddleware>());
+
+        var middlewareTask = middleware.Invoke(httpContext);
+
+        await writtenHeaders.Task;
+
+        Assert.Contains(TestSink.Writes, w => w.Message.Contains("StatusCode: 101"));
+        Assert.Contains(TestSink.Writes, w => w.Message.Contains("Connection: Upgrade"));
+
+        letBodyFinish.SetResult();
+
+        await middlewareTask;
+    }
+
+    [Fact]
+    public async Task UpgradeToWebSocketLogsWrittenOnlyOnce()
+    {
+        var options = CreateOptionsAccessor();
+        options.CurrentValue.LoggingFields = HttpLoggingFields.ResponsePropertiesAndHeaders;
```

Comment:
Adding some variation to ensure the different combinations work.
```suggestion
        options.CurrentValue.LoggingFields = HttpLoggingFields.ResponseHeaders;
```